src/main/java/br/edu/utfpr/reclamaguarapuava/controller/ProblemResource.java

```diff
@@ -0,0 +1,174 @@
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        Problem problem = new Problem(dto);
+        try {
+            problemService.save(problem);
+        } catch (Exception e) {
+            response.addError("Houve um erro ao persistir os seus dados.");
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        dto = new ProblemDTO(problem);
+        response.setData(dto);
+
+        return ResponseEntity.ok(response);
+    }
+
+    @PutMapping(value = "/{id}")
+    public ResponseEntity<?> update(@PathVariable Long id, @Valid @RequestBody ProblemDTO dto, BindingResult result) {
+        Response<ProblemDTO> response = new Response<>();
+
+        if (result.hasErrors()) {
+            response.setErrors(result);
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        Optional<Problem> problemOptional = problemService.findById(id);
+        if (!problemOptional.isPresent()) {
+            response.addError("Problema não encontrado.");
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        Problem problem = problemOptional.get();
+
+        if (!problem.getName().equals(dto.getName())) {
+            Optional<Problem> problem1 = problemService.findByName(dto.getName());
+            if (problem1.isPresent() && problem1.get().getDescription().equals(dto.getDescription())) {
+                response.addError("Nome e descrição sendo usados em outro problema");
+                return ResponseEntity.badRequest().body(response);
+            }
+        }
+
+        problem.update(dto);
+        problem = problemService.save(problem);
+
+        dto = new ProblemDTO(problem);
+        response.setData(dto);
+
+        return ResponseEntity.ok(response);
+    }
+
+    @DeleteMapping(value = "/{id}")
+    public ResponseEntity<Response<String>> delete(@PathVariable Long id) {
+        log.info("Removendo problema com id {}", id);
+
+        Response<String> response = new Response<>();
+
+        Optional<Problem> problemOptional = problemService.findById(id);
+
+        if (!occurrenceService.findByProblem(id).isEmpty()) {
+            response.addError("Não é possível apagar problema com ocorrências associadas");
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        if (!problemOptional.isPresent()) {
+            response.addError("Erro ao apagar problema com o id " + id);
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        try {
+            this.problemService.deleteById(id);
+        } catch (ObjectDeletedException e) {
+            response.addError("Erro ao apagar problema");
+            return ResponseEntity.badRequest().body(response);
+        }
+
+        return ResponseEntity.ok(response);
+    }
+
+    @GetMapping(value = "/paginacao")
+    public ResponseEntity<Response<Page<ProblemDTO>>> findAllPaginationWithPage(
```

Comment:
este metodo de paginacao nao é preciso pelo seguinte caso:
no seu findAll() basta por como um dos parâmetros do tipo Pageable e pronto ja estará com TODOS as variáveis de paginação, dai basta apenas na hr q for chamar o service.findall() já passar a paginação q o próprio Spring se responsabiliza, tem exemplo nos sources q mandei no slack do grupo

Reply:
outra coisa, seguindo a arquitetura rest a url para o findar, nao deve se chamar paginação, nem todos, nem nada, é apenas o nome do recurso:
`/api/admin/categorias`
então onde esta:
`@GetMapping(value = "/paginacao")`
deve ficar como:
`@GetMapping`